Make Flips ND compatible and fix ND ScaleKeepAspect, ScaleFixed and PinOrigin

itan1 opened the pull request:
- Updated some of the projective transform functions such that they are ND compatible (ScaleKeepAspect, ScaleFixed, PinOrigin)
- Added a generic ND compatible FlipDim and made FlipX FlipY and FlipZ aliases
- Added those transformations to the 3D test case in test/projective/affine.jl

The transforms `Rotate` and `Reflect` are still only 2D compatible.

closes #74

itan1 committed:
Fix getprojection() methods for ND case for ScaleKeepAspect, ScaleFixed and PinOrigin. Also add ScaleFixed and Zoom to the 3D test to catch the bug.

itan1 committed:
make centered() ND compatible, add Flips for ND data and test them in the 3D test

lorenzoh commented on src/projective/affine.jl:
```diff
@@ -178,24 +178,45 @@ end
 
 
+"""
+    FlipDim{N}()
```
Think this should be
```suggestion
    FlipDim{N}(dim)
```

itan1 replied:
fixed

lorenzoh commented on src/projective/affine.jl:
```diff
@@ -178,24 +178,45 @@ end
+    FlipDim{N}(dim) where {N} = 1 <= dim <= N ? new{N}(dim) : error("invalid dimension")
+end
+# 2D images use (r, c) = (y, x) convention
+FlipX(N) = FlipDim{N}(N==2 ? 2 : 1)
+FlipY(N) = FlipDim{N}(N==2 ? 1 : 2)
+FlipZ(N) = FlipDim{N}(3)
```
Can we make this so you call `FlipX{N}()`? The way it is I find it somewhat confusing that for `FlipDim` the type variable is the dimension of the data, where as for `FlipXYZ` the dimension of the data is the argument.

itan1 replied:
yes makes sense, I changed it accordingly

lorenzoh reviewed (dismissed):
Thanks for the contribution!

This looks good generally, I've left some comments. It seems that this only touches Flip, not Zoom like the title implies, right?

Other than the minor suggested changes, could you add some tests that test that these transformations behave correctly in 3D? Some sanity checks, i.e.

- `FlipX{3}()` applied to a `(10, 10, 1)` image should be the same as `FlipX{2}()` to a  `(10, 10)` image (dropping the last dimension). Same for `FlipY`
- `FlipZ{3}()` applied to a `(10, 10, 1)` image should be an identity operation

If you're testing this on 3D images, I would also be interested in some visual examples as an additional check. Let me know if you something is unclear!

itan1 committed:
Change N from argument to parametric type for FlipX, FlipY and FlipZ constructors, to match the FlipDim definition

itan1 committed:
add 2D and 3D test for flips: flip dimension size of 1 means identity operation

itan1 commented:
Thank you for the feedback :)

Yes you're right, `Zoom` is not affected. Only Flip is added and I fixed a small bug in `ScaleFixed`, `ScaleKeepAspect` and `PinOrigin`.

I just added 2 tests for images where the flip dimension has size 1, but I ran into a `BoundsError` in Interpolations.jl. Applying the flip transform fails, both for 2D and 3D. I could not reproduce the error in the REPL but apparently [during tests Julia always uses bounds checking](https://discourse.julialang.org/t/test-fails-when-run-through-pkg-test-but-not-when-run-interactively/67352/3). That explains why it only occurs in the tests. I haven't found the exact reason yet but will have another look.

itan1 committed:
add tests for all 2D and 3D flip indices, add test for double flip = identity

itan1 commented:
The arrays with singleton dimensions did not work well with Interpolations.jl, they give bounds errors during testing.

Instead of the singleton tests I checked the flipped image pixel values by reverse indexing and added a test to check that double flipping results in the original image.

CarloLucibello commented:
@lorenzoh can this be merged?

paulnovo committed:
Merge branch 'master' into make-flip-and-zoom-nd-compatible

paulnovo committed:
Fix Flips documentation and tests errors caused by master merge

This adds missing documentation for the Flip transformations, and adds
FlipDim to the reference page. Also, fixes the "Big Pipeline/3D" unit
test that was failing due to not being able to compose two
ComposedProjectiveTransforms.

paulnovo commented:
I merged master into this branch and updated a few things in [this branch](https://github.com/paulnovo/DataAugmentation.jl/tree/make-flip-and-zoom-nd-compatible). I can help with any additional needed changes to get this merged. If everything looks good, should I create a PR into this `itan1` branch? Or a new PR into FluxML/Data augmentation.jl?

paulnovo reviewed (approved):
I think this is ready to merge, @lorenzoh any objections?
